Filter while recursing

bugfood opened the pull request:
*Issue #*
#1138
Probably #1117

*Description of changes:*
This fleshes out https://github.com/aws/aws-cli/pull/2105 and hopefully is suitable for inclusion. The patch from @clarete is unmodified except for resolution of a trivial merge conflict.

More description is available in https://github.com/aws/aws-cli/pull/2105 ; my contributions are to:
* Add a minor optimization.
* Add a unit test.
* Simplify surrounding unit tests.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bugfood commented:
Thanks @IhostVlad for the approval. Is there anything else I can/should do in order to get this merged?

Thanks,
Corey

clarete commented:
Thanks for working on this, @bugfood!  I wonder if @kyleknap could take a look at this and maybe help us getting this merged? 🙏

bugfood commented:
Thanks @clarete for the original patch! I set out with the goal to write such a patch and found it already pretty much done.

bugfood commented:
@IhostVlad can you please merge this, or can you please tell me whom I should ask?

bugfood commented:
@kyleknap @vz10 sorry to pester, but I see that you have each merged some pull requests in recent months. Can one of you please check this PR and merge it?

Right now, one of our production backup jobs is broken because of an unreadable file in a snapshot dir that is excluded. With this PR, the excluded dir would not even be traversed and we would not be having that problem. It would be really nice to get this change upstream.

If there's anything else I need to do, please let me know. Thanks.

bugfood commented:
One of our sync jobs is failing again due to this.

I opened case number 9628762541 with our corporate AWS account a few weeks ago to see if they can get any traction on getting the PR merged. No news yet.

tim-finnigan commented:
Hi @bugfood thanks for the PR and sorry for the wait. Our team just put out a recent proposal in https://github.com/aws/aws-cli/pull/6828 detailing improvements to the contribution process. We are working through open PRs and have determined this PR to be in the Implementation stage ([described here](https://github.com/kdaily/aws-cli/blob/contrib-guide-proposal/proposals/contribution-guide.md#implementation-stage-1)). Here is some initial feedback going through the PR:

* We would want all of the filtering to be done in the `FileGenerator`. This will reduce the logical complexity of potentially applying the filter twice depending on which method gets called. So this means that the listing of S3 objects will need to apply filters along with removing filters in our iterator chain.
* We need more functional tests around using `--exclude` and `--include`. This would require end-to-end tests where we run an entire CLI command without making network calls. Right now we really only have a handful of unit and integration tests for this and so if we make a larger change there is not too much of a safety net that we did not introduce regressions to the commands. I think to start, we would port the integration and unit tests over to functional tests and need to decide if there are any cases that we are missing. We'd also want to add new functional tests for behavior that we do not want to get such as no longer warning on files if they are excluded.

We are tracking this PR [here](https://github.com/orgs/aws/projects/21/views/2) on our project board and it can be set to to the review stage once new changes are ready to be reviewed.

bugfood commented:
Thanks @tim-finnigan. I can try to work on this next week, but I'd need some clarification on your feedback, or else I can wait until the surrounding plans firm up, if necessary.

Specifically:
> We would want all of the filtering to be done in the `FileGenerator`. ...

I'm definitely hazy on the details after all this time, but I checked the PR and the change does indeed move the filtering into the `FileGenerator`, unless I'm missing something; the rest of the changes are just to pass the filter parameter downward to the `FileGenerator`. Or does your comment relate to a bigger-picture notion of other filter handling?

> We need more functional tests around using `--exclude` and `--include`. ...

That sounds reasonable, but I don't think I'm in a good position to handle that. I can look at the testing next week, but what you describe sounds like more of a larger refactor and fleshing out of the testing system. Do you consider that a prerequisite for making any changes like the one in this PR? Mostly I tried to take @clarete 's PR and ensure there wasn't a regression in testing coverage. On the subject of the contribution process in general, it's tricky for a first-time contributor like me to work on improvements that are peripheral to the specific intent of the PR, at least without some very specific guidance.

Thanks,
Corey

tim-finnigan commented:
Thanks @bugfood for following up. The filtering changes I mentioned do require a larger refactoring effort. Currently filtering takes place both in the `FileGenerator` as well as after it. But we want to remove the filters [here](https://github.com/aws/aws-cli/blob/18cfa4daa73098e426cdd0bb2f5630fa4243346e/awscli/customizations/s3/subcommands.py#L1028-L1029) from the command dictionary (for the `cp` and `mv` commands as well). And we want to add the file filter [here](https://github.com/aws/aws-cli/pull/6828) in `list_objects`. This will make it so filtering is only happening within the `FileGenerator`.

The addition of functional tests for `--include` and `--exclude` could be split into a separate PR but that would need to be merged first.

We appreciate you contributing your efforts on the AWS CLI and understand that the steps can be ambiguous, especially as a first time contributor. If you have any feedback on how we can improve the experience for you, please consider leaving a comment on our proposal for the contributing process here (https://github.com/aws/aws-cli/pull/6828). Thanks!

bugfood commented:
I have uploaded a new PR rebased against current `master`.

Changes from the previous are:

1. I added a functional test for the interaction between `--exclude` and `--include`, and that uncovered a bug in the original PR: an earlier exclude of a directory would mask an include of a subdirectory, even if the include was specified later.
2. To fix the bug, the best way I could think to handle this would be to auto-generate include rules for the parent directories of any includes. This way, we still traverse into those parent directories in order to find the included directories. This is included as a separate preparatory patch before @clarete's patch. You can experiment locally: remove the auto-generate patch to see the new include/exclude test fail.
3. I added a couple more patches for trivial cleanup of nearby code.
4. I added a test to check that excluded files are not examined. This checks for a failure similar to that in #1117. You can experiment locally: remove @clarete's patch to see this test fail.

I did not include any refactoring changes to the `Filegenerator`. I do not know if you wanted that to be done here, but I do think that it should be out of scope for this PR. The changes introduced here should not hinder such a later refactoring, nor do they make the situation any worse without the refactor having been done. I think bringing the refactor into scope would be an unnecessary burden and delay for this contribution.

bugfood commented:
Do PRs no longer get tested by travis-ci? I don't see a link for that. I have run the test suite locally for Python 3.9.10, but can't guarantee other versions pass.

kdaily commented:
Thanks @bugfood! Since you were a first time contributor, the GitHub Action for testing didn't run. I just approved it, so you should be good to go to see the testing results.

bugfood commented:
@kdaily thanks.

Hmm, windows trouble... I'll look into it.

bugfood commented:
I pushed some changes intended to fix the test runs on windows.

@kdaily can you please approve the test run or make the action be automatic for me?

https://github.com/aws/aws-cli/actions/runs/2168429541

![image](https://user-images.githubusercontent.com/2976480/163435737-3ae09eb2-1e8d-4526-a498-b95f645935af.png)

bugfood commented:
Ah, hmm. The test suite ran against my own account, it seems; I got an email about the results.

https://github.com/bugfood/aws-cli/actions/runs/2168429422

That isn't linked to this PR, though, which is why I was confused.

bugfood commented:
The test suite now passes:
https://github.com/bugfood/aws-cli/actions/runs/2168554422

This is obsolete and can be cancelled:
https://github.com/aws/aws-cli/actions/runs/2168429541

This needs to be run:
https://github.com/aws/aws-cli/actions/runs/2168554777

justindho commented:
@bugfood Thanks for making the changes! The team has done another pass on the PR and had some questions/comments.

1. Is there a specific reason that additional filtering logic was added to [filters.py](https://github.com/aws/aws-cli/pull/5425/files#diff-4ccac8d4b48491d94c15470ee4c84dbd240c796c1872db8ac13cf32194177fd1R104)? As @tim-finnigan mentioned in a [previous comment](https://github.com/aws/aws-cli/pull/5425#issuecomment-1090602181), we'd like to remove the filters from the `sync`, `cp`, `rm`, and `mv` [subcommands](https://github.com/aws/aws-cli/pull/5425/files#diff-46a47bc72f6ef2ce6c174a439ca750bb2a63ad9b4151d0b45e66c74448d194bfL1028) and keep filtering to one location (in the `FileGenerator`'s [`list_objects`](https://github.com/aws/aws-cli/blob/18cfa4daa73098e426cdd0bb2f5630fa4243346e/awscli/customizations/s3/filegenerator.py#L306) method) to reduce complexity. We recognize that this is a bit of a refactor, but the team would still like to see the changes done here.
2. Could you rebase your changes off of the `develop` branch? It looks like the changes were rebased off of `master`, which introduced a lot of commits to this PR.
3. The tests you've added look pretty good. Just one minor comment: could you also add another assertion in [text_include_exclude](https://github.com/aws/aws-cli/pull/5425/files#diff-7bc71137597f698d3265a46c9b947008aeeb8a8584262bb9c60c6cbc2316fb39R103) to verify that no operations were called with excluded paths (i.e., `self.assertNotIn(path, excluded)`)?

bugfood commented:
Thanks for the review.

I rebased first, then I added a couple commits, but I need to do more work on that later, so please don't re-review yet.

I added the new `assertNotIn`; no problem there.

I had earlier added the code to filters.py because the surrounding code is involved in the compilation of user-supplied filters into the actual filters used internally. Since I needed to auto-insert some patterns at pattern-compilation time, that seemed like a natural place. I'm not arguing that it _must_ be that way, but I don't know what would be better.

For the filtering location, I'm not entirely sure what you are asking for and I may need clarification later, though I did spot a deficiency to address first.

bugfood commented:
@justindho can you please check the PR now?

The last two commits are work done today; the previous commits are unchanged.

I _think_ this does what you meant with respect to moving the filtering into FileGenerator. I had pushed back on this earlier, but now I realize that the PR was making the filter handing inconsistent, so it is good to take care of this now. At least, if I understand correctly.

kyleknap reviewed (commented):
Hi @bugfood! I'll be the primary reviewer for the PR and will be reviewing any updates to the PR until it gets merged.

With regards to this comment:
> I think this does what you meant with respect to moving the filtering into FileGenerator. I had pushed back on this earlier, but now I realize that the PR was making the filter handing inconsistent, so it is good to take care of this now. At least, if I understand correctly.

Yes that is what we were referring to. The update here looks great.

Otherwise, I mainly looked through the implementation and all of my comments were related to the implementation. I'll probably have some other smaller comments in a subsequent review, but I want to make sure we are in a good spot with regards to the approach to the implementation first. Thanks again for seeing this through!

bugfood commented on awscli/customizations/s3/filters.py:
```diff
@@ -101,6 +101,20 @@ def _full_path_patterns(self, original_patterns, rootdir):
         for pattern in original_patterns:
             full_patterns.append(
                 (pattern[0], os.path.join(rootdir, pattern[1])))
+            # When including paths deep in the directory tree, auto-include the
```
Thanks. I have to focus on another project for the time being, but I will circle back to this later for the technical changes requested.

> I'm still not sure I understand the case this logic is optimizing for after reading this comment: https://github.com/aws/aws-cli/pull/5425#issuecomment-1125343591 and the code. Would it be possible to illustrate the use/edge case with a sample directory structure and CLI commands?

The use case is fairly simple:
```
aws s3 sync dir s3://bucket/dir '--exclude excluded/*' --include 'excluded/included/*'
```

If we make the FileGenerator no longer recurse into excluded directories, then `excluded/included` would never be examined and any files therein would have no chance to be included. What is needed is a specific include filter for `excluded/included`. We could make the user do it:
```
aws s3 sync dir s3://bucket/dir '--exclude excluded/*' --include excluded/included --include 'excluded/included/*'
```

...but that seems onerous and would break current behavior. In my PR, this is now verified by the commit with description `add a functional test for "aws s3 sync" include/exclude`.

The best solution I thought of was to auto-insert the necessary filter, since we know the existing filtering logic does what we need. I wanted to keep the special-case code to a minimum.

kyleknap replied:
Thanks for the detailed examples that was really helpful in understanding it! In general with this PR, we definitely need to make sure we do not break current behavior (i.e. given any set of filters + file set, the CLI will continue to include/exclude the same set of files after this change with no needing to adjust the filters provided at the command line).

While it looks like the coded that got added does handle the missing edge case that you illustrated, I am concerned that we are going to be missing other edge cases after seeing the edge case you raised. For example, suppose we had this directory layout:
```
$ tree .
.
├── directory
│   ├── another-dir
│   │   ├── exclude-me.txt
│   │   └── include-me.py
│   ├── exclude-me.txt
│   └── include-me.py
├── exclude-me.txt
└── include-me.py
```
If we only wanted to upload the `.py` files, a user would likely run a command like this:
```
$ aws s3 cp . s3://mybucket --recursive --exclude '*' --include '*.py' --dryrun
```
This would result in all of the `.py` files being uploaded:
```
(dryrun) upload: directory/another-dir/include-me.py to s3://mybucket/directory/another-dir/include-me.py
(dryrun) upload: directory/include-me.py to s3://mybucket/directory/include-me.py
(dryrun) upload: ./include-me.py to s3://mybucket/include-me.py
```
However, with the approach we are currently taking, we would still not traversing the `directory` and `another-dir` to get all of the .py files:
```
$ aws s3 cp . s3://mybucket --recursive --exclude '*' --include '*.py' --dryrun
(dryrun) upload: ./include-me.py to s3://mybucket/include-me.py
```
In general, I'm not optimistic the original approach of just applying the filters to the directory before traversing it is going to work. In terms of options to pursue, I think we either:

* Need to expand the logic of whether we traverse a directory to skip traversing a directory only if it is impossible for a file inside of it to be included based on the filters. I'm not sure if there is a straightforward way of doing this as we would need to account for all supported pattern symbols: https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3/index.html#use-of-exclude-and-include-filters

* Scope the optimization of not traversing directories to simple set of exclude/include patterns (e.g. we only apply this optimization to only filters that have only `--exclude` flags and/or apply the optimizations if there are no special characters used in the filter (e.g. `*`, `?`, `[]`, `[!]`) 

Thoughts? I want to think about this a bit more to see if I have any better suggestions, but I at least wanted to put this realization down now to start the discussion.

bugfood replied:
Hmm, that is rough.

Your first idea seems best, but I don't know how to do that either.

A third idea would be to give the user control, either by a global option `--no-traverse-excludes` or a separate argument `--hard-exclude <path>` (to replace `--exclude <path>`). This isn't ideal, since it would be best if the program does the right thing automatically, but it may be more straightforward to implement and would not have unexpected consequences for current uses.

Thebutlersguilty3 commented:
Corey Hickey Friday, May 27, 5:34 PM
--hard-exclude


Sent from my Metro by T-Mobile 5G Device
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Corey Hickey ***@***.***>
Sent: Friday, May 27, 2022 5:34:01 PM
To: aws/aws-cli ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [aws/aws-cli] Filter while recursing (#5425)


@bugfood commented on this pull request.

________________________________

In awscli/customizations/s3/filters.py<https://github.com/aws/aws-cli/pull/5425#discussion_r883991435>:

> @@ -101,6 +101,20 @@ def _full_path_patterns(self, original_patterns, rootdir):
         for pattern in original_patterns:
             full_patterns.append(
                 (pattern[0], os.path.join(rootdir, pattern[1])))
+            # When including paths deep in the directory tree, auto-include the


Hmm, that is rough.

Your first idea seems best, but I don't know how to do that either.

A third idea would be to give the user control, either by a global option --no-traverse-excludes or a separate argument --hard-exclude <path> (to replace --exclude <path>). This isn't ideal, since it would be best if the program does the right thing automatically, but it may be more straightforward to implement and would not have unexpected consequences for current uses.

—
Reply to this email directly, view it on GitHub<https://github.com/aws/aws-cli/pull/5425#discussion_r883991435>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AWLW4QLSPAPBDERAZF63KMDVME5UTANCNFSM4PK6QO4Q>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

tim-finnigan commented:
Upon discussing with the team, this PR is not prioritized at this time but we are setting it to Draft to continue tracking the proposed changes. Per [this earlier comment](https://github.com/aws/aws-cli/pull/5425#discussion_r883991435) and discussion there are still some design implementation options that must be considered for a path forward here.

tim-finnigan commented:
Checking in — we are going to close this as not prioritized at this time. There has not been any discussion regarding the changes in a couple of years, and further research is required into the best implementation here. But we are continuing to track the open related issues (https://github.com/aws/aws-cli/issues/1138 and https://github.com/aws/aws-cli/issues/1117)
